File: src/components/dashboard/reportRuntimeStructure.test.js
import assert from 'node:assert/strict';
import fs from 'node:fs';
import { resolveReportRuntimeCompositeOwnership } from './reportRuntimeStructure.js';
const fixture = JSON.parse(fs.readFileSync(new URL('../../reporting/fixtures/report-runtime-structure-conformance.v1.json', import.meta.url)));
for (const scenario of fixture.cases) {
  const actual = resolveReportRuntimeCompositeOwnership(scenario.blocks);
  assert.deepEqual(Object.fromEntries(actual.parentById), scenario.parents, scenario.name);
  assert.deepEqual(scenario.blocks.filter((block) => !actual.childBlockIdSet.has(block.id)).map((block) => block.id), scenario.roots, scenario.name);
}
console.log('Report runtime structure conformance ✓');

File: android/sdk/src/test/java/com/viant/forgeandroid/runtime/ReportRuntimeStructureTest.kt
package com.viant.forgeandroid.runtime

import java.io.File
import kotlinx.serialization.json.*
import kotlin.test.*

class ReportRuntimeStructureTest {
    @Test fun sharedCompositeOwnershipContract() {
        val root = generateSequence(File(System.getProperty("user.dir"))) { it.parentFile }
            .first { File(it, "src/reporting/fixtures/report-runtime-structure-conformance.v1.json").exists() }
        val fixture = Json.parseToJsonElement(File(root, "src/reporting/fixtures/report-runtime-structure-conformance.v1.json").readText()).jsonObject
        for (scenario in fixture["cases"]!!.jsonArray.map { it.jsonObject }) {
            val blocks = dashboardReportRuntimeBlocks(scenario["blocks"]!!.jsonArray)
            assertEquals(scenario["parents"], JsonObject(blocks.mapNotNull { block -> block.compositeParentId?.let { block.id to JsonPrimitive(it) } }.toMap()))
            assertEquals(scenario["roots"], JsonArray(blocks.filter { it.compositeParentId == null }.map { JsonPrimitive(it.id) }))
            blocks.forEach { block -> assertTrue(block.children.all { it.compositeParentId == block.id }) }
        }
    }
}
